### PR TITLE
expression, util:  Adjust Time value checking rule according to SQLMode

### PR DESCRIPTION
**Pull request by spongedu**
According to https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_no_zero_date, sql mode value `NO_ZERO_DATE` and `NO_ZERO_IN_DATE` works separately before Mysql  5.7.4 and after  Mysql 5.7.8，and the behavior in `SELECT` statement diffs from that in `INSERT` and `UPDATE`.  This PR adjusts the behavior of TiDB time check according to sql_mode. Please take a look @zz-jason

### CODE REVIEW

**Comment by shenli** [comment]
@spongedu CI failed.

**Comment by zz-jason** [comment]
to #4373

File: util/types/time.go [hunk 1]
```diff
@@ -35,6 +37,11 @@ var (
 	ErrInvalidYear       = errors.New("invalid year")
 )
 
+// ErrIncorrectDatetimeValue generate an 'Incorrect datetime value: XXX' error
+func ErrIncorrectDatetimeValue(v string) error {
```
Comment (zz-jason) [inline util/types/time.go]: you can use `types.ErrIncorrectDatetimeValue` defined in file "util/types/time.go"

**Comment by zz-jason** [comment]
ci failed:

```
FAIL: integration_test.go:1019: testIntegrationSuite.TestTimeBuiltin
integration_test.go:1056:
    c.Assert(terror.ErrorEqual(err, types.ErrInvalidTimeFormat), IsTrue)
... obtained bool = false
```

**Comment by zz-jason** [comment]
@spongedu any update ?

**Comment by zhexuany** [comment]
@spongedu friendly ping. Any updates?

File: expression/builtin_time.go [hunk 2]
```diff
@@ -975,13 +975,13 @@ type builtinDayOfMonthSig struct {
 // evalInt evals a builtinDayOfMonthSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_dayofmonth
 func (b *builtinDayOfMonthSig) evalInt(row types.Row) (int64, bool, error) {
+	oldMode := b.ctx.GetSessionVars().SQLMode
+	b.ctx.GetSessionVars().SQLMode = oldMode & (^mysql.ModeNoZeroInDate)
```
Comment (spongedu) [inline expression/builtin_time.go]: According to https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_dayofmonth,  date value like `2008-00-00` is valid for builtin function `dayofmonth`, so sql mode `NO_ZERO_IN_DATE` has to be canceld when calculating. It's somewhat ugly way here, I think the thorough solution is to refine the time check & sql mode handling system, it's something big. @zz-jason

Comment (zz-jason) [reply expression/builtin_time.go]: in mysql:
```sql
MySQL > select dayOfMonth('2017-00-00');
+--------------------------+
| dayOfMonth('2017-00-00') |
+--------------------------+
|                        0 |
+--------------------------+
1 row in set (0.01 sec)

MySQL > select dayOfMonth('0000-00-00');
+--------------------------+
| dayOfMonth('0000-00-00') |
+--------------------------+
|                     NULL |
+--------------------------+
1 row in set, 1 warning (0.01 sec)

MySQL > show warnings;
+---------+------+----------------------------------------+
| Level   | Code | Message                                |
+---------+------+----------------------------------------+
| Warning | 1292 | Incorrect datetime value: '0000-00-00' |
+---------+------+----------------------------------------+
1 row in set (0.01 sec)
```

Comment (spongedu) [reply expression/builtin_time.go]: @zz-jason  yes. That's what I do: cancelling `NO_ZERO_IN_DATE`, reserve `NO_ZERO_DATE`.  Your examples acts the same with my change. PTAL

File: expression/integration_test.go [hunk 3]
```diff
@@ -1163,7 +1163,7 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
 	result.Check(testkit.Rows("<nil>"))
 	result = tk.MustQuery("select timediff('-34 12:00:00', '2014-1-2 12:00:00');")
 	result.Check(testkit.Rows("<nil>"))
-	result = tk.MustQuery("select timediff('2014-1-2 12:00:00', '12:00:00');")
```
[the remark this answers is not in the document]
Comment (spongedu) [reply expression/integration_test.go]: @zz-jason it's deleted by mistake. I'll fix it

**Comment by zz-jason** [comment]
@spongedu any update ?

File: expression/builtin_time_test.go [hunk 4]
```diff
@@ -170,8 +170,8 @@ func (s *testEvaluatorSuite) TestDate(c *C) {
 		YearWeek   interface{}
 	}{
 		{nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil},
-		{"0000-00-00 00:00:00", nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil},
-		{"0000-00-00", nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil},
+		{"0000-00-00 00:00:00", nil, nil, nil, 0, nil, nil, nil, nil, nil, nil, nil},
```
Comment (tiancaiamao) [inline expression/builtin_time_test.go]: why change test?

File: expression/builtin_time_test.go [hunk 5]
```diff
@@ -370,20 +370,21 @@ func (s *testEvaluatorSuite) TestDayOfMonth(c *C) {
 	defer testleak.AfterTest(c)()
 	sc := s.ctx.GetSessionVars().StmtCtx
 	sc.IgnoreZeroInDate = true
```
Comment (tiancaiamao) [inline expression/builtin_time_test.go]: IgnoreZeroInDate is true, what's the problem with the test and why you change it?

File: expression/builtin_time_test.go [hunk 6]
```diff
@@ -370,20 +370,21 @@ func (s *testEvaluatorSuite) TestDayOfMonth(c *C) {
...
-		{"0000-00-00 12:12:12", 0, false, false},
+		{1, false, false},
+		{0, false, false},
+		{0, false, false},
+		{0, false, false},
+		{0, false, false},
 	}
-	for _, t := range cases {
-		f, err := newFunctionForTest(s.ctx, ast.DayOfMonth, s.primitiveValsToConstants([]interface{}{t.args})...)
+	for i := 1; i < len(cases); i++ {
```
Comment (tiancaiamao) [inline expression/builtin_time_test.go]: why `i := 1`

File: types/time.go [hunk 7]
```diff
@@ -1338,6 +1348,41 @@ func checkDateType(t TimeInternal, allowZeroInDate bool) error {
...
 
+func checkDateTypeInModify(t TimeInternal, sc *variable.StatementContext) error {
+	year, month, day := t.Year(), t.Month(), t.Day()
+	strict, noZeroDate, noZeroInDate, ignoreErr := true, true, true, false
+	if sc != nil && (sc.InInsertStmt || sc.InUpdateOrDeleteStmt) {
+		m := sc.SQLMode
+		strict, noZeroDate, noZeroInDate, ignoreErr = m.HasStrictMode(), m.HasNoZeroDateMode(), m.HasNoZeroInDateMode(), sc.IgnoreErr
+	}
+	if year == 0 && month == 0 && day == 0 {
+		if !noZeroDate {
```
Comment (tiancaiamao) [inline types/time.go]: Double negative make the logic hard to read.

File: types/time.go [hunk 8]
```diff
@@ -1338,6 +1348,41 @@ func checkDateType(t TimeInternal, allowZeroInDate bool) error {
...
+			return nil
+		} else if !strict || ignoreErr {
+			sc.AppendWarning(ErrWarnDataOutOfRange)
+		}
+		return nil
```
Comment (tiancaiamao) [inline types/time.go]: If `noZeroDate` is true, and in strict mode, insert 0000-00-00 returns no error here?

File: expression/builtin_time_test.go (hunk 6 above)
Comment (spongedu) [reply expression/builtin_time_test.go]: @tiancaiamao by mistake. Will fix

File: expression/builtin_time_test.go (hunk 5 above)
Comment (spongedu) [reply expression/builtin_time_test.go]: @tiancaiamao the changes are to test different combination of sql_mode `NO_ZERO_DATE` and `STRICT_TRANS_TABLES`. `IgnoreZeroInDate` actually don't represent these sql_modes

File: types/time.go (hunk 8 above)
Comment (spongedu) [reply types/time.go]: no, there should be an error. There seems still something wrong. I'll take a look

File: types/time.go (hunk 7 above)
Comment (spongedu) [reply types/time.go]: ok. I'll try to refine it.

**Comment by zz-jason** [comment]
@spongedu any update? And plz merge master && resolve conflicts.

**Comment by zhexuany** [comment]
@spongedu just a friendly ping. This PR is staled.

**Comment by zz-jason** [comment]
I'm going to close this PR due to it's outdated. Please feel free to reopen it if you are going to working on this again.

**Pull request closed without merging**